### ISSUE TITLE
OSVR Server/client apps hang on shutdown

### ISSUE THREAD

**Issue by mars979** [vrpn/vrpn#108]
When I'm running OSVR Server or client application like AnalogCallback_cpp.exe from cmd window and I exit out with "Ctrl-C", then instead of returning, it hangs without exiting.
Based on the stack trace below it looks like the thread is stuck waiting to become unblocked.

Server stack trace
[External Code]    
osvrConnectiond.dll!vrpn_Semaphore::p() Line 240
osvrConnectiond.dll!vrpn::SemaphoreGuard::lock() Line 446
osvrConnectiond.dll!vrpn::SemaphoreGuard::SemaphoreGuard(vrpn_Semaphore & sem) Line 427
osvrConnectiond.dll!vrpn_ConnectionManager::deleteConnection(vrpn_Connection * c) Line 1481
osvrConnectiond.dll!vrpn_Connection_IP::~vrpn_Connection_IP() Line 6220
[External Code]    
>    osvrConnectiond.dll!vrpn_ConnectionManager::~vrpn_ConnectionManager() Line 1440
     [External Code] 
   
Autos:
Name                   Value                  Type
hSemaphore        0x00000160	void *
this                       0x0f587638         {osvrConnectiond.dll!vrpn_ConnectionManager manager} {cResources=1 hSemaphore=0x00000160 }	vrpn_Semaphore *

Client stack trace
 	[External Code]	
>	osvrClientd.dll!vrpn_Semaphore::p() Line 240
 	osvrClientd.dll!vrpn::SemaphoreGuard::lock() Line 446
 	osvrClientd.dll!vrpn::SemaphoreGuard::SemaphoreGuard(vrpn_Semaphore & sem) Line 427
 	osvrClientd.dll!vrpn_ConnectionManager::deleteConnection(vrpn_Connection * c) Line 1481
 	osvrClientd.dll!vrpn_Connection_IP::~vrpn_Connection_IP() Line 6220
 	[External Code]	
 	osvrClientd.dll!vrpn_ConnectionManager::~vrpn_ConnectionManager() Line 1437
 	[External Code]	

  
Autos:
Name                   Value                  Type
hSemaphore        0x00000170        void *
this	                     0x0fbf5518  {osvrClientd.dll!vrpn_ConnectionManager manager} {cResources=1 hSemaphore=0x00000170 }	vrpn_Semaphore *

**Comment by rpavlik** [comment]
Got this happening in an OSVR branch I've pushed called vrpn-instrumentation, where I added ETW custom events. (The `JointClientKitWithAsyncInterface` test will trigger the bug pretty reliably for me and requires no server to be running.) The time sequence of events that is happening leading up to the hang is:

```
vrpn_ConnectionManager::~vrpn_ConnectionManager()
SemaphoreGuard: attempting to lock
SemaphoreGuard: locked
vrpn_ConnectionManager is deleting a connection
vrpn_ConnectionManager::instance()
SemaphoreGuard: attempting to lock
SemaphoreGuard: locked
SemaphoreGuard: unlocked
SemaphoreGuard: attempting to lock
```

Sadly I forgot to build with debug symbols on, so I can't tell you where those exactly take place, but it looks like the problem happens in the destructor of the connection manager, as it tries to delete a connection.

![screenshot 096](https://cloud.githubusercontent.com/assets/61129/9888562/6210a0e4-5bbb-11e5-85b1-9945eb7ec96c.png)


Interestingly, while the reports are all from one thread (even if I 'zoom out' to the full execution time), I can trigger the issue most reliably if I'm using an OSVR "async" plugin (like the video tracker or this dummy eye tracker) that gets its own thread.
